addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java

```diff
@@ -0,0 +1,723 @@
+     * @param pCmd the command string from the item.
+     * @return {@link TivoStatusData} object.
+     */
+    public TivoStatusData cmdTivoSend(String pCmd) {
+
+        boolean oC = false; // Tracks if this step caused a connection to be opened
+        TivoStatusData tmpStatus = null;
+        int retryCount = 2;
+
+        pausePolling = true;
+
+        // Open a connection if one does not already exist
+        if (!connTestSocket()) {
+            connTivoConnectRetry(true);
+            oC = true;
+        }
+
+        tmpStatus = cmdTivoSendRetry(pCmd, retryCount);
+
+        // Will close the connection after operation if we are not keeping one open.
+        if (!tivoConfigData.isCfgKeepConnOpen() && oC && !tivoStatusData.getChScan()) {
+            connTivoConnectRetry(false);
+        }
+        pausePolling = false;
+        return tmpStatus;
+
+    }
+
+    /**
+     * {@link cmdTivoSendRetry} sets the TiVo status, executes the specified number of retries.
+     *
+     * @param pCmd the command string from the item.
+     * @param retryCount the retry count.
+     * @return {@link TivoStatusData} object.
+     */
+    private TivoStatusData cmdTivoSendRetry(String pCmd, int retryCount) {
+
+        logger.info(" TiVo '{}' - sending command: '{}' (retry @ '{}')", tivoConfigData.getCfgIdentifier(), pCmd,
+                retryCount);
+        int l = 1;
+
+        if (!connTestSocket()) {
+            logger.error(" TiVo '{}' - called setTivoStatus but not connected!", tivoConfigData.getCfgIdentifier());
+            return new TivoStatusData(false, -1, "NOT_CONNECTED", false, 0);
+        }
+
+        try {
+            // Handle special keyboard "repeat" commands
+            if (pCmd.contains("*")) {
+                l = Integer.parseInt(pCmd.substring(pCmd.indexOf("*") + 1));
+                pCmd = pCmd.substring(0, pCmd.indexOf("*"));
+                logger.info(" TiVo '{}' - repeating command: '{}' repeating '{}' times",
+                        tivoConfigData.getCfgIdentifier(), pCmd, l);
+            }
+            for (int i = 1; i <= l; i++) {
+                tivoIOSendCommand.println(pCmd.toString() + "\r");
+                if (tivoIOSendCommand.checkError()) {
+
+                    logger.error(" TiVo '{}' - called setTivoStatus and encountered an IO error",
+                            tivoConfigData.getCfgIdentifier(), tivoSocket.isConnected(), tivoSocket.isClosed());
+                    connTivoConnectRetry(false);
+                    connTivoConnectRetry(true);
+
+                    if (retryCount <= 0) {
+                        // We have failed at life... Give up.
+                        return new TivoStatusData(false, -1, "CONNECTION_RETRIES_EXHAUSTED", false, 0);
+
+                    } else {
+                        doNappTime();
+                        return cmdTivoSendRetry(pCmd, retryCount - 1);
+                    }
+                }
+            }
+
+        } catch (Exception e) {
```

Comment:
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception